hmmlearn/hmm.py

```diff
@@ -520,134 +525,294 @@ class GMMHMM(_BaseHMM):
+                             (np.ones((self.n_components, 1)) * self.n_mix))
+
+        if 'm' in self.init_params or not hasattr(self, "means_"):
+            self.means_ = np.zeros((self.n_components, self.n_mix,
+                                    self.n_features))
+            for i, kmeans in enumerate(kmeanses):
+                self.means_[i] = kmeans.cluster_centers_
+
+        if 'c' in self.init_params or not hasattr(self, "covars_"):
+            cv = np.cov(X.T) + self.min_covar * np.eye(self.n_features)
+            if not cv.shape:
+                cv.shape = (1, 1)
+
+            if self.covariance_type == 'tied':
+                self.covars_ = np.zeros((self.n_components,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'full':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'diag':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features))
+                self.covars_[:] = np.diag(cv)
+            elif self.covariance_type == 'spherical':
+                self.covars_ = np.zeros((self.n_components, self.n_mix))
+                self.covars_[:] = cv.mean()
+
+    def _check(self):
+        super(GMMHMM, self)._check()
+
+        if not hasattr(self, "n_features"):
+            self.n_features = self.means_.shape[2]
+
+        # Checking covariance type
+        if self.covariance_type not in COVARIANCE_TYPES:
+            raise ValueError("covariance_type must be one of {0}"
+                             .format(COVARIANCE_TYPES))
+
+        self.weights_ = np.array(self.weights_)
+        # Checking mixture weights' shape
+        if self.weights_.shape != (self.n_components, self.n_mix):
+            raise ValueError("mixture weights must have shape "
+                             "(n_components, n_mix), "
+                             "actual shape: {}".format(self.weights_.shape))
+
+        # Checking mixture weights' mathematical correctness
+        if not np.allclose(np.sum(self.weights_, axis=1),
+                           np.ones(self.n_components)):
+            raise ValueError("mixture weights must sum up to 1")
+
+        # Checking means' shape
+        self.means_ = np.array(self.means_)
+        if self.means_.shape != (self.n_components, self.n_mix,
+                                 self.n_features):
+            raise ValueError("mixture means must have shape "
+                             "(n_components, n_mix, n_features), "
+                             "actual shape: {}".format(self.means_.shape))
+
+        # Checking covariances' shape
+        self.covars_ = np.array(self.covars_)
```

Comment:
Functions can be used to improve readability. Take a look at the code. It's long and complex, so you've put in a comment communicating the intent to the reader. However, you could've just as well used a function:

```python
validate_covars_shape(...)
validate_covars_consistency(...)

# Or just validate_covars(...)
```

So, instead of obfuscating the code, the function can make it easier for the reader to get the "bigger" picture of what's going on. Assuming the function has aa good name of course :)